lib/galaxy/webapps/galaxy/api/tool_data.py

```diff
@@ -1,21 +1,143 @@
+
+
+@cbv(router)
+class FastAPIToolData:
+    tool_data_manager: ToolDataManager = Depends(get_tool_data_manager)
+
+    @router.get(
+        '/api/tool_data',
+        summary="Lists all available data tables",
+        response_description="A list with details on individual data tables.",
+        dependencies=[
+            AdminUserRequired
+        ],
+    )
+    async def index(self) -> ToolDataEntryList:
+        """Get the list of all available data tables."""
+        return self.tool_data_manager.index()
+
+    @router.get(
+        '/api/tool_data/{name}',
+        summary="Get details of a given data table",
+        response_description="A description of the given data table and its content",
+        dependencies=[
+            AdminUserRequired
+        ],
+    )
+    async def show(self, name: str = ToolDataTableName) -> ToolDataDetails:
+        """Get details of a given tool data table."""
+        return self.tool_data_manager.show(name)
+
+    @router.get(
+        '/api/tool_data/{name}/reload',
+        summary="Reloads a tool data table",
+        response_description="A description of the reloaded data table and its content",
+        dependencies=[
+            AdminUserRequired
+        ],
+    )
+    async def reload(self, name: str = ToolDataTableName) -> ToolDataDetails:
+        """Reloads a data table and return its details."""
+        return self.tool_data_manager.reload(name)
+
+    @router.get(
+        '/api/tool_data/{table_name}/fields/{field_name}',
+        summary="Get information about a particular field in a tool data table",
+        response_description="Information about a data table field",
+        dependencies=[
+            AdminUserRequired
+        ],
+    )
+    async def show_field(
+        self,
+        table_name: str = ToolDataTableName,
+        field_name: str = ToolDataTableFieldName,
+    ) -> ToolDataField:
+        """Reloads a data table and return its details."""
+        return self.tool_data_manager.show_field(table_name, field_name)
+
+    @router.get(
+        '/api/tool_data/{table_name}/fields/{field_name}/files/{file_name}',
+        summary="Get information about a particular field in a tool data table",
+        response_description="Information about a data table field",
+        response_class=FileResponse,
+        dependencies=[
+            AdminUserRequired
+        ],
+    )
+    async def download_field_file(
+        self,
+        table_name: str = ToolDataTableName,
+        field_name: str = ToolDataTableFieldName,
+        file_name: str = Path(
+            ...,  # Mark this field as required
+            title="File name",
+            description="The name of a file associated with this data table field",
+        ),
+    ):
+        """Download a file associated with the data table field."""
+        path = self.tool_data_manager.get_field_file_path(table_name, field_name, file_name)
+        return FileResponse(str(path))
```

Comment:
I think we'll want to do what we do in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/framework/base.py#L484 somehow.
Maybe doing that as a middleware would be the best approach ?

Reply:
@mvdbeek 
I have been playing around with something like:
````python
def add_send_file_middleware(app: FastAPI, gx_config=Depends(?)) -> None:
    @app.middleware("http")
    async def add_send_file_header(request: Request, call_next) -> Response:
        response = await call_next(request)
        if isinstance(response, StreamingResponse):
            if gx_config.nginx_x_accel_redirect_base:
                response.headers['X-Accel-Redirect'] = path_to_file
            elif gx_config.apache_xsendfile:
                response.headers['X-Sendfile'] = path_to_file
        return response
````
But I had some troubles figuring out how to get the config properly. Maybe when we merge #11180 we can look into it again? I will put it in my TODO list, promise :)

Reply:
The test data for these tools is relatively small so it is probably less important this works with this controller, but when we get to the datasets controller and such we will definitely need that working I think.